File: Leetcode 0094 Binary Tree Inorder Traversal [medium] [tree]/Binary Tree Inorder Traversal [iterative].js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var inorderTraversal = function(root) {
    let inorder = [];
    let stack = [];
    
    if (root === null) {
        return inorder;
    }
    
    let cur = root;
    while (stack.length !== 0 || cur !== null) {
        while (cur !== null) {
            stack.push(cur);
            cur = cur.left;
        }
        
        if (stack.length !== 0) {
            cur = stack.pop();
            inorder.push(cur.val);
            cur = cur.right;
        }
    }
    
    return inorder;
};

File: Leetcode 0094 Binary Tree Inorder Traversal [medium] [tree]/Binary Tree Inorder Traversal [iterative].java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> inorder = new ArrayList<>();
        Stack<TreeNode> stack = new Stack<>();
        
        if (root == null) {
            return inorder;
        }
        
        TreeNode cur = root;
        while (!stack.isEmpty() || cur != null) {
            while (cur != null) {
                stack.push(cur);
                cur = cur.left;
            }
            
            if (!stack.isEmpty()) {
                cur = stack.pop();
                inorder.add(cur.val);
                cur = cur.right;
            }
        }
        
        return inorder;
    }
}